Split configuration of TLSv1.3 ciphers from older ciphers

mattcaswell opened the pull request:
With the current mechanism, old cipher strings that used to work in 1.1.0,
may inadvertently disable all TLSv1.3 ciphersuites causing connections to
fail. This is confusing for users.

In reality TLSv1.3 are quite different to older ciphers. They are much
simpler and there are only a small number of them so, arguably, they don't
need the same level of control that the older ciphers have.

This change splits the configuration of TLSv1.3 ciphers from older ones.
By default the TLSv1.3 ciphers are on, so you cannot inadvertently disable
them through your existing config.

Fixes #5359

A few things to note about this (which may or may not be controversial):

- I have taken the opportunity to replace the OpenSSL specific names for the ciphersuites with the RFC standard names. The only reason to have OpenSSL specific ones was to make them a bit more consistent with the rest of the ciphersuite names. Now that they are treated differently anyway that reason seems less compelling (if it ever was).

- This PR also changes the default ordering of the TLSv1.3 ciphersuites. Previously the TLSv1.3 ciphersuites were mixed in with the rest of the ciphersuites in the order "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256". Now they are at the top of the ciphersuites list (on the basis that we would rather use these TLSv1.3 ciphersuites than older ones) and in the order "TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384". This is on the basis that AES-256 seems overkill for the default choice (it also has the side benefit that the default hash will now be SHA-256 which is one step towards easing the PSK transition as discussed in #5378). I'm not sure which we would prefer as the top choice out of the AES vs chacha20 ciphersuites. Note this is slightly inconsistent with the default ordering of the older ciphersuites which are sorted by strength (so AES-256 is preferred over AES-128). Since we are handling the two lists differently, I kind of think that is ok.

- I have added new functions `SSL_CTX_set_ciphersuites()`/`SSL_set_ciphersuites()` for configuring the TLSv1.3 specific ciphersuites. There is also a new "Ciphersuites" SSL_CONF parameter, as well as a new "-ciphersuites" command line option for s_server/s_client/ciphers.

- We possibly need `SSL_CTX_get0_ciphersuites()` and `SSL_get0_ciphersuites()` too.

- Under the hood the TLSv1.3 ciphersuites still get added into the same cipher_list as all the others. This is visible to an application if it calls `SSL_get_ciphers()` et al.
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [x] tests are added or updated

mattcaswell commented:
Pushed a fix for the Travis failure. Still investigating the AppVeyor failure.

richsalz commented:
I think using "ciphersuites" is confusing, maybe "tls13ciphers" ?
Why does apps/s_time.c have a getenv(SSL_CIPHERS) ?

mattcaswell commented:
> I think using "ciphersuites" is confusing, maybe "tls13ciphers" ?

I was deliberately trying to use something which *doesn't* include the string tls1.3 in it. That will live with us forever even when it is no longer relevant.

> Why does apps/s_time.c have a getenv(SSL_CIPHERS) ?

No idea. It doesn't seem documented. Although I should at least add an option to s_time to set the TLSv1.3 ciphersuites.

richsalz commented:
Yes, add the flag to s_time and remove the option when you do so.
As for using/notusing tls1.3 in the name, I'm curious what others think.

t-j-h commented:
Don't use tls1.3 in the name - as we want this to be used in future.

kroeckx commented:
I prefer that we default AES-256 over AES-128

kroeckx commented:
I'm wondering if there will be any interaction between this new setting and older TLS protocols.

FdaSilvaYY commented on include/openssl/tls1.h:
```diff
@@ -1118,16 +1118,6 @@ __owur int SSL_check_chain(SSL *s, X509 *x, EVP_PKEY *pk, STACK_OF(X509) *chain)
-# define TLS1_3_TXT_AES_256_GCM_SHA384                     "TLS13-AES-256-GCM-SHA384"
-# define TLS1_3_TXT_CHACHA20_POLY1305_SHA256               "TLS13-CHACHA20-POLY1305-SHA256"
-# define TLS1_3_TXT_AES_128_CCM_SHA256                     "TLS13-AES-128-CCM-SHA256"
 # define TLS1_3_TXT_AES_128_CCM_8_SHA256                   "TLS13-AES-128-CCM-8-SHA256"
```
Why do you keep this last one  ? ;)

mattcaswell replied:
Oops. How did I miss that one? Fixed now.

FdaSilvaYY commented on ssl/ssl_conf.c:
```diff
@@ -263,6 +263,16 @@ static int cmd_CipherString(SSL_CONF_CTX *cctx, const char *value)
 
+static int cmd_Ciphersuites(SSL_CONF_CTX *cctx, const char *value)
+{
+    int rv = 1;
```
style: blank line after decl.

mattcaswell replied:
Fixed

mattcaswell commented:
>  Yes, add the flag to s_time and remove the option when you do so.

I added the flag. I didn't remove the environment variable thing. If people are relying on that it could break things - which seems unnecessary.

mattcaswell commented:
> I prefer that we default AES-256 over AES-128

What do other people think about this? BTW I realised that I had inadvertently also changed the ordering of the chacha ciphersuite vs the AES-128 one. I've now swapped those two around. I've currently kept aes-128/chacha being negotiated in favour of AES-256 for now until I get further feedback.

mattcaswell commented:
> I'm wondering if there will be any interaction between this new setting and older TLS protocols.

I'm not sure what sort of thing you're worried about here?

richsalz commented:
I do not think we should be setting the preference; it goes against our philosophy. We can also expect that some users -- e.g., FIPS -- will need to turn off some ciphers -- e,.g., ChaCha -- and we should allow them to otherwise use TLS 1.3

mattcaswell commented:
> I do not think we should be setting the preference; it goes against our philosophy.

I'm not sure I understand what you are trying to say? What preference are you talking about here?

> We can also expect that some users -- e.g., FIPS -- will need to turn off some ciphers -- e,.g., ChaCha -- and we should allow them to otherwise use TLS 1.3

This PR enables people to turn off what ever ciphers they want to?

richsalz commented:
Maybe I misunderstand, I did skim the code but perhaps not read closely enough.  Comments like "kept aes 128 over 256" implies there is a specific ordering of ciphers.  No?

mattcaswell commented:
> Maybe I misunderstand, I did skim the code but perhaps not read closely enough. Comments like "kept aes 128 over 256" implies there is a specific ordering of ciphers. No?

Well yes. In the code before this PR (and before TLSv1.3) we order our ciphersuites by preference. This is the DEFAULT cipher list. It is a requirement that we know what that order is for us to construct or process a ClientHello. This PR just separates out the configuration of TLSv1.3 ciphersuites from TLSv1.2 and below. We still need to order the ciphersuites as we did before this PR.

Prior to this PR the ordering was:

"TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256"

As part of this PR I changed the order to:

"TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384"

On the basis that AES256 as a default first preference seems overkill. Kurt thinks we should keep the ordering the same as it was before. It is that issue I am asking for comment on.

richsalz commented:
Can I, as an application, configure which ciphers I want to use, and in which order?

mattcaswell commented:
> Can I, as an application, configure which ciphers I want to use, and in which order?

Yes, absolutely. We're just talking about the default here if an application doesn't specify anything different.

kroeckx commented:
On Mon, Feb 19, 2018 at 01:48:26PM +0000, mattcaswell wrote:
> > I'm wondering if there will be any interaction between this new setting and older TLS protocols.
> 
> I'm not sure what sort of thing you're worried about here?

I'm wondering if in the new configuration you say "AES", or
"AES-256-GCM-SHA384", that you're going to enable the TLS 1.2
ciphers that match that or not, or if this is going to be a TLS
1.3 setting only.

Do we also have (and need) settings for the key exchange and
authentication?

mattcaswell commented:
> I'm wondering if in the new configuration you say "AES", or
"AES-256-GCM-SHA384", that you're going to enable the TLS 1.2
ciphers that match that or not, or if this is going to be a TLS
1.3 setting only.

No. The new configuration settings for TLSv1.3 are a simple list of the allowed ciphersuites. It does not have all the functionality of the TLSv1.2 cipher list (it's not needed - there are currently only 5 permissible values). It is not possible to specify "AES" in the list for TLSv1.3 because it is not a ciphersuite name.

kroeckx commented:
My question is if it's useful to actually support it, that we split the current setting in 3 settings and that it's still general purpose. We could then deprecate the old one. You could then for instance have 3DES in it, and it would select the 3DES ciphers for older versions but since we don't have any ciphers for TLS 1.3 with 3DES it wouldn't do anything for TLS 1.3.

t-j-h commented:
I think we should keep the ordering as it was - ordered by strength (as that is what we have as our DEFAULT) for TLSv1.2 - why would we elected to pick a different order for TLSv1.3 - i.e. not by strength - don't we want to selected the "strongest" available as our default?

i.e. I think I agree with @kroeckx

kroeckx commented:
I'm not sure at all how my proposal to split it up in 3 and work
for both TLS 1.2 and TLS 1.2 is supposed to work.

mattcaswell commented:
> I'm not sure at all how my proposal to split it up in 3 and work
for both TLS 1.2 and TLS 1.2 is supposed to work.

No, I don't know either, although it might be interesting to explore it. However, I think it is probably dealing with a separate problem to the one we are addressing here.

I think your proposal only really works if applications switch to your new 3-way configuration, and ditch any existing cipher_list configuration - otherwise how do you resolve conflicts between the two configuration approaches?

That's fine, and you could have the two things running in parallel, as long as applications only use one of them. But the problem this PR is trying to address is the backwards compatibility issue where applications have set a cipher_list that excludes the possibility of TLSv1.3 ciphersuites and expect it to just work, i.e. we need a solution which allows TLSv1.3 to work *in the presence* of an existing cipher list configuration.

So - in other words, I think your proposal is interesting, and might be worth exploring, but doesn't impact this PR.

mattcaswell commented:
Fix up commit pushed to change the order back to preferring AES-256 over the AES-128/CHACHA20

kroeckx commented:
I guess the way to split it is ignoring the new ones if the old one is set.

mattcaswell commented:
> I guess the way to split it is ignoring the new ones if the old one is set.

Right. But that doesn't help with the problem at hand: legacy client cipher strings can disable all TLSv1.3 ciphersuites. So I think we still need this PR. Your proposal still has merit but is orthogonal to this change.

mattcaswell commented:
Pushed a new commit to try and keep the CIs happy. Lets see if it worked. But, anyway, I'm taking this out of WIP. Please review!

mattcaswell commented:
Ping @openssl/committers. This would be nice to get in for next week's alpha if possible.

mattcaswell commented:
Rebased this due to conflicts with master.

Ping? Still looking for a reviewer for this.

t-j-h commented:
You've addressed all of my comments - I've got nothing further to add. It looks good to me.
Leave it non-approved for others to comment - but I will come back and approve if no one else does - given I've re-read through the code again.

t8m commented:
It would be nice if @SECLEVEL still applied to the tls1.3 ciphersuite selection.

mattcaswell commented:
> It would be nice if @seclevel still applied to the tls1.3 ciphersuite selection.

I'm not sure what your use case is here. If you use "@seclevel" in the TLSv1.2 cipher string then it applies universally (so would apply equally to TLSv1.3 ciphersuites). But I'm not sure what practical impact that really has?

tmshort commented on apps/s_time.c:
```diff
@@ -59,17 +59,20 @@ static const size_t fmt_http_get_cmd_size = sizeof(fmt_http_get_cmd) - 2;
      "Where to connect as post:port (default is " SSL_CONNECT_NAME ")"},
-    {"cipher", OPT_CIPHER, 's', "Cipher to use, see 'openssl ciphers'"},
+    {"cipher", OPT_CIPHER, 's',
+     "TLSv1.3 and below cipher list to use, see 'openssl ciphers'"},
```
This description is inconsistent with the `-cipher` option in apps/apps.h; which says "TLSv1.2 and below..."

richsalz replied:
I dislike a reference to a manpage, if that's even what it is.  That should be implicit in every command; see the manpage for full(er) information.  Yes, I probably did this text.  But what does the single-quoted text *mean*?

mattcaswell replied:
> This description is inconsistent with the -cipher option in apps/apps.h; which says "Specify TLSv1.2 and below..."

It should say TLSv1.2 and below.

> I dislike a reference to a manpage, if that's even what it is. That should be implicit in every command; see the manpage for full(er) information. Yes, I probably did this text. But what does the single-quoted text mean?

I think it refers to the output from running that command - but its not that clear. I can remove that bit of the text.

mattcaswell replied:
Done.

tmshort commented:
What happens if a user puts TLSv1.3 ciphers into the "TLSv1.2 and below" list? (For those who've been using the older versions, or are unaware of the new API?)
And subsequently, how does this affect applications compiled against 1.1.0, that pull in the 1.1.1 shared object library, and hope/believe they'll get TLSv1.3 support by merely changing the cipher string?

mattcaswell commented:
> What happens if a user puts TLSv1.3 ciphers into the "TLSv1.2 and below" list? (For those who've been using the older versions, or are unaware of the new API?)

The same as what happens if you put any unrecognised string into the cipher string - it is ignored.

> And subsequently, how does this affect applications compiled against 1.1.0, that pull in the 1.1.1 shared object library, and hope/believe they'll get TLSv1.3 support by merely changing the cipher string?

There is no impact on the configuration of TLSv1.3 ciphersuites, no matter what the contents of your TLSv1.2 and below cipher string. That's kind of the point. Before this PR a 1.1.0 cipher string used in 1.1.1 can inadvertently disable all TLSv1.3 ciphersuites. After this PR the old cipher string configuration will never have any impact on the availability of TLSv1.3 ciphersuites. Unless you explicitly take other steps, you will always have the default TLSv1.3 ciphersuites available.

tmshort commented:
I guess my concern is the possibility of losing the capability for a 1.1.0-compiled app to be able to update the dynamically-linked library and get TLSv1.3 for free... Previous to this PR, it would require that TLSv1.3 ciphers be added to the cipher string, now there's a separate default TLSv1.3 cipher string. So, an application would get TLSv1.3 for free.

With this PR, there's no need to do anything special to get TLSv1.3 if the 1.1.1 library is dynamically loaded. However, it may mean getting TLSv1.3 with unexpected ciphers (e.g. ChaCha20) if that cipher were not included in the original cipher string. Not sure if this is a bad thing or a reasonable thing. 

Which also leads to an application getting TLSv1.3 even if they didn't want it; since the TLSv1.3 ciphers are defined. This can lead to a change in behavior of applications (compiled against 1.1.0) that previously only used TLSv1.2 if it dynamically loads the 1.1.1 library. This could lead to compatibility issues. Not sure if this is a bad thing or a good thing.

Before this PR, all that behavior could be controlled via the single cipher string configuration.

mattcaswell commented:
> I guess my concern is the possibility of losing the capability for a 1.1.0-compiled app to be able to update the dynamically-linked library and get TLSv1.3 for free... Previous to this PR, it would require that TLSv1.3 ciphers be added to the cipher string, now there's a separate default TLSv1.3 cipher string. So, an application would get TLSv1.3 for free.
>
> With this PR, there's no need to do anything special to get TLSv1.3 if the 1.1.1 library is dynamically loaded. However, it may mean getting TLSv1.3 with unexpected ciphers (e.g. ChaCha20) if that cipher were not included in the original cipher string. Not sure if this is a bad thing or a reasonable thing.

As soon as we enable TLSv1.3 by default we are in a catch-22. Either:

1) The old cipherstring mechanism is used for both TLSv1.2 and TLSv1.3 and then
    1a) That means that any connection to a TLSv1.3 capable peer will fail for some configurations
    or
    1b) To avoid 1a we force all connections to fail until the configuration is fixed
    or
    1c) We detect the condition and silently disable TLSv1.3, meaning that lots of applications that believe they have enabled TLSv1.3 actually haven't and have no easy way of finding out that they haven't

or

2) We separate out the TLSv1.3 ciphersuite configuration from TLSv1.2 and below. The argument here is that TLSv1.3 ciphersuites are really very different to TLSv1.2 and below ciphersuites anyway. This could mean applications may actually have some ciphersuites enabled that they weren't expecting

Of the above we started off with 1a. We decided that was bad because you get spurious occasional failures which are hard to diagnose. So then we moved to 1b. This has proven to also be bad - we've had a number of issues raised where people have not understood the cause of the error. This PR moves to (2) (this is the same approach as Boring takes...except you can't actually configure the TLSv1.3 ciphersuites at all in Boring).


> Which also leads to an application getting TLSv1.3 even if they didn't want it; since the TLSv1.3 ciphers are defined. This can lead to a change in behavior of applications (compiled against 1.1.0) that previously only used TLSv1.2 if it dynamically loads the 1.1.1 library. This could lead to compatibility issues. Not sure if this is a bad thing or a good thing.

Either we have TLSv1.3 defaulted to on, and risk that some applications didn't really want it. Or we have it defaulted to off, which I think would significantly slow the adoption of TLSv1.3 and would generally be a bad thing. If its on by default then there is no way to avoid the potential for some compatibility issues. By definition if we default it to on, then sometimes we are going to negotiate TLSv1.3 - which works very differently to TLSv1.2.

> Before this PR, all that behavior could be controlled via the single cipher string configuration.

The argument is that they really are different things. TLSv1.2 ciphersuites are not compatible with TLSv1.3 ciphersuites and vice versa. They work in a different way, so in a lot of ways it makes sense to separate out the configuration.

t8m commented:
@mattcaswell If @SECLEVEL in the normal cipher string applies also to limit the tls-1.3 ciphersuites (f.e. to limit only to >128 bit ciphersuites with seclevel 4), then it is fine.

tmshort commented:
@mattcaswell It sounds as though it's been well-thought out, thank you.

mattcaswell commented:
I made a minor tweak to the help text in speed based on the feedback above. I've also rebased to resolve a conflict with master. Ping @t-j-h.

vdukhovni commented on ssl/ssl_lib.c:
```diff
@@ -2520,8 +2528,9 @@ int SSL_CTX_set_cipher_list(SSL_CTX *ctx, const char *str)
-                                &ctx->cipher_list_by_id, str, ctx->cert);
+    sk = ssl_create_cipher_list(ctx->method, ctx->tls13_ciphersuites,
+                                &ctx->cipher_list, &ctx->cipher_list_by_id, str,
+                                ctx->cert);
```
Can you explain what's happening here?  Is this parsing a single string for both sets of ciphers?

vdukhovni replied:
Never mind, I think I see that`tls13_ciphersuites` is just passed through unaltered by the value of "str" and prepended to the cipher stack as-is.  What does this about the order of calls that configure TLS13 vs. TLS12 ciphers? Does this mean that the TLS13 configuration (if performed) has to happen first?

vdukhovni commented on ssl/ssl_lib.c:
```diff
@@ -2555,6 +2565,99 @@ int SSL_set_cipher_list(SSL *s, const char *str)
+
+    if (ret && s->cipher_list != NULL) {
+        /* We already have a cipher_list, so we need to update it */
+        return update_cipher_list(s->cipher_list, s->tls13_ciphersuites);
```
I see, if the order is reversed, you're patching the combined stack... C'est la vie?

mattcaswell replied:
Right!

vdukhovni commented on ssl/ssl_ciph.c:
```diff
@@ -1469,6 +1471,15 @@ STACK_OF(SSL_CIPHER) *ssl_create_cipher_list(const SSL_METHOD *ssl_method, STACK
+            return NULL;
+        }
+    }
+
```
I see the TLS 1.3 ciphers go on top.  How unusual do you think it will be for users to build OpenSSL 1.1.1 with RC4 and TLS 1.0 still enabled?  Some rather dated TLS 1.0 libraries only see the first 64 ciphers in the client handshake, and only support TLS 1.0 with RC4 and 3DES.  When these get pushed further down the list, they become invisible to these (Microsoft Windows 2003 Server most likely) implementations. So one might argue that the TLS 1.3 ciphers should go last (since the protocol is chosen first, this won't matter to real TLS 1.3 peers).  On the other hand, do we just say good-riddance to interoperability with 15-year old TLS 1.0 stacks with only ciphers we no longer enable by default?

vdukhovni replied:
Of course users who desperately want interop, can tweak the TLS 1.2 cipher list to make sure that RC4 (at least for Exchange2003 3DES is broken, and only RC4 works) is above slot 62 (rather than 65 as before)

mattcaswell replied:
> I see the TLS 1.3 ciphers go on top. How unusual do you think it will be for users to build OpenSSL 1.1.1 with RC4 and TLS 1.0 still enabled? Some rather dated TLS 1.0 libraries only see the first 64 ciphers in the client handshake, and only support TLS 1.0 with RC4 and 3DES.

By default we're only sending through 3 TLSv1.3 ciphersuites, so that still leaves the next 61 ciphersuites for a match to be found for these old (buggy) libraries. Personally I don't think we should worry about this.

vdukhovni commented:
A broader question.  In TLS 1.2 we have not only specific cipher names, but also names for classes of ciphers sharing a common property.  Some of the common properties pertain not to the bulk cipher but to the key agreement protocol and its signature.  Thus we have kECDHE, kDHE, aRSA, aECDSA, ...

The TLS 1.3 cipherlist does not appear to offer any control over the key agreement parameters. Should it? And if so how?  And if not, why not?  Say, for example one wanted to avoid the ECDHE groups and use only the FFDHE groups or vice-versa?

Or accept only certificates, and not ECDSA certificates?  And do we support Ed25519 / Ed448 certificates yet?  If so are these just generically ECDSA?  Or do they get their own distinct names?  What are we doing with the "cert" key slots in the SSL structures, last I looked there were slots for DSA, RSA and ECDSA, but the design does not easily accomodate multiple flavours of ECDSA.

Indeed TLS 1.3 adds a new `signature_algorithms_cert` extension, which is much more specific about the supported algorithms:
```
      enum {
          /* RSASSA-PKCS1-v1_5 algorithms */
          rsa_pkcs1_sha256(0x0401),
          rsa_pkcs1_sha384(0x0501),
          rsa_pkcs1_sha512(0x0601),

          /* ECDSA algorithms */
          ecdsa_secp256r1_sha256(0x0403),
          ecdsa_secp384r1_sha384(0x0503),
          ecdsa_secp521r1_sha512(0x0603),

          /* RSASSA-PSS algorithms with public key OID rsaEncryption */
          rsa_pss_rsae_sha256(0x0804),
          rsa_pss_rsae_sha384(0x0805),
          rsa_pss_rsae_sha512(0x0806),

          /* EdDSA algorithms */
          ed25519(0x0807),
          ed448(0x0808),

          /* RSASSA-PSS algorithms with public key OID RSASSA-PSS */
          rsa_pss_pss_sha256(0x0809),
          rsa_pss_pss_sha384(0x080a),
          rsa_pss_pss_sha512(0x080b),

          /* Legacy algorithms */
          rsa_pkcs1_sha1(0x0201),
          ecdsa_sha1(0x0203),

          /* Reserved Code Points */
          private_use(0xFE00..0xFFFF),
          (0xFFFF)
      } SignatureScheme;
```
Which in principle makes it possible to provision separate P-256, P-384, Ed25519, ... certificates and negotiate which to use. But IIRC we not at all ready to support such fine-grained configurations. So perhaps further evolution of TLS 1.3 cipherlists is gated on a design for fine-grained control of the certificate types?

Anyway, it is my impression that work remains to be done in this space, and this interface is somewhat preliminary.

mattcaswell commented:
> The TLS 1.3 cipherlist does not appear to offer any control over the key agreement parameters. Should it? And if so how? And if not, why not? Say, for example one wanted to avoid the ECDHE groups and use only the FFDHE groups or vice-versa?

No, I don't think it should. In TLSv1.3 none of these things are part of the chosen ciphersuite any more in the spec (ciphersuites are quite different in TLSv1.3). They are all controlled through other extensions. Anything you want to be able to configure along those lines should be configurable by other means. We do not currently support FFDHE groups for TLSv1.3 at all. If you want to have control over the specific groups that could be used then we have the ability to configure the contents of the supported groups extension (e.g. see docs for SSL_CTX_set1_groups_list(), or the "Groups" config option for SSL_CONF).

> Or accept only certificates, and not ECDSA certificates?

Did you mean to write accept only RSA certicates, and not ECDSA certificates? If so then this is configurable through sig algs, e.g. see docs for SSL_CTX_set1_sigalgs_list() or the "SignatureAlgorithms" config option for SSL_CONF.

> And do we support Ed25519 / Ed448 certificates yet?

Yes.

> If so are these just generically ECDSA? Or do they get their own distinct names?

We follow the sig algs extension so they have their own names.

> What are we doing with the "cert" key slots in the SSL structures, last I looked there were slots for DSA, RSA and ECDSA, but the design does not easily accomodate multiple flavours of ECDSA.

We now have additional slots for Ed25519 and Ed448. For traditional ECDSA certs you can currently only have one, i.e. you couldn't have a P-256 ECDSA cert and a P-384 ECDSA cert at the same time (but we've never supported that - that's not changed for TLSv1.3)

> Indeed TLS 1.3 adds a new signature_algorithms_cert extension, which is much more specific about the supported algorithms:
> Which in principle makes it possible to provision separate P-256, P-384, Ed25519, ... certificates and negotiate which to use. But IIRC we not at all ready to support such fine-grained configurations.

signature_algorithms_cert enables you to specify which signatures you are willing to accept *in certificates*. It says nothing about the public keys held within those certs. For that you still use signature_algorithms. As a server it is perfectly possible to set up multiple certificates, e.g. you could have an RSA cert, an ECDSA cert, an Ed25519 cert and an Ed448 cert. As a client you can specify which ones you are willing to accept by configuring sig algs.

We have full support for the signature_algorithms extension and partial support for signature_algorithms_cert (we can receive it but not send it).

> So perhaps further evolution of TLS 1.3 cipherlists is gated on a design for fine-grained control of the certificate types?

I don't believe the two are connected. Ciphersuites in TLSv1.3 are really quite simple, and have nothing to do with certificate selection.

vdukhovni commented:
> @mattcaswell No, I don't think it should. In TLSv1.3 none of these things are part of the chosen ciphersuite any more in the spec (ciphersuites are quite different in TLSv1.3). They are all controlled through other extensions.

Matt, while I am aware of and agree that the specification has switched from a list of all-in-one cipher code points to just bulk cipher + PRF, with handshake parameters negotiated separately, I don't think that this (necessarily or ideally) means that the user-interface should be partitioned in the same way.

Most applications will present just one configuration element to the user to configure the relevant settings, and will not to expose lots of separate settings that will likely confuse the user.

Prior to TLS 1.3, the relevant settings could all be configured via the cipherlist.  I think that limiting the syntax to just the bulk ciphers is likely a poor user interface.  You might compare with GNU-TLS, where the "priority string" offers extensive control over the cryptographic settings: https://gnutls.org/manual/html_node/Priority-Strings.html

I strongly feel that users should be able to select not only the bulk ciphers, but also the acceptable peer signature algorithms, and supported groups.  All via the new TLS 1.3-style cipherstring.  The underlying C API can present multiple special-purpose interfaces for changing one thing at a time, but the ciphersuite string that applications expose to users should I think be more flexible.

richsalz commented:
Can't those extensions be added later?

vdukhovni commented:
A good user interface translates the user's mental model of the task to the internal model of the program, rather than just expose and instrument the internal model, and ask the user to adapt to that.

richsalz commented:
I don't disagree with that sentiment.  I just question how important it is that we provide full fine-grain control of TLS 1.3 in this initial release.

mattcaswell commented:
> I don't think that this (necessarily or ideally) means that the user-interface should be partitioned in the same way.

> Most applications will present just one configuration element to the user to configure the relevant settings, and will not to expose lots of separate settings that will likely confuse the user.

I totally and strongly disagree. I think it is *far* more confusing to push configuration of lots of different concepts into a single string. I'm not sure what evidence the "most applications will present just one configuration element to the user" comes from - but if so it is hardly surprising if there has been only one configuration element for these things to date. Previously it could all (mostly) be managed by the cipher string. There is absolutely no reason in my mind to continue that model into a future where it just does not fit well.

mattcaswell commented:
>  I just question how important it is that we provide full fine-grain control of TLS 1.3 in this initial release.

Just to point out that we *have* fine-grain control of these things for TLSv1.3. Just not in one string like @vdukhovni wants. Doing that would be a bad mis-feature in my mind.

vdukhovni commented:
> @richsalz I don't disagree with that sentiment. I just question how important it is that we provide full fine-grain control of TLS 1.3 in this initial release.

Looking upthread we see:
> @vdukhovni Anyway, it is my impression that work remains to be done in this space, and this interface is somewhat preliminary.

So I was mainly trying to say that I think we're not done yet.  The question of when a more complete interface should be added is separate.  If 1.1.1 were an LTS release candidate, then I'd say that we need to flesh out the ciphersuite syntax now.  If this a TLS 1.3 preview, with LTS support in 1.2 or similar, then this can wait a bit, but early adopters of 1.1.1 will find that they're losing some control over the cryptographic parameters unless they add new user-interface elements to expose the additional separate controls.

If different applications expose different interfaces, users will not see a consistent approach across the application ecosystem.  So there is a downside to deferring the interface design too long...

mattcaswell commented:
> A good user interface translates the user's mental model of the task to the internal model of the program, 

Most user's mental model today will be based on how TLSv1.2 works. TLSv1.3 does not work like that. It would be wrong for us to attempt to fit TLSv1.3 into a TLSv1.2 mental model. That would ultimately result in more confusion. As users become more accustomed to how TLSv1.3 works they will expect to be able to configure it in a way that matches that. Or putting it another way our user's mental model will change as they begin to understand TLSv1.3. Our task is to provide an interface that matches what that model *will be* not what it is *now*.

mattcaswell commented:
> early adopters of 1.1.1 will find that they're losing some control over the cryptographic parameters unless they add new user-interface elements to expose the additional separate controls.

TLSv1.3 is *new*. It works differently. There are multiple things that have changed that apps will need to modify to take advantage of the new features (not just around configuration).  Applications will have to adapt regardless. That won't happen over night. It would be very wrong of us to try and contort things in a way that might help a bit now but leave us with a bad interface in the long run.

> there is a downside to deferring the interface design too long...

I do not see this as deferring the interface design at all. We *have* an interface that provides a good granularity of control. It works a bit differently to how it worked before. That's not unreasonable. TLSv1.3 itself works differently to how things worked before.

vdukhovni commented:
> @mattcaswell Most user's mental model today will be based on how TLSv1.2 works. TLSv1.3 does not work like that. It would be wrong for us to attempt to fit TLSv1.3 into a TLSv1.2 mental model.

I don't think users should need to understand that internally OpenSSL has separate interfaces for configuring ciphers vs. key exchange groups vs. signature algorithms, ...  And most won't know much about TLS 1.3.

They'll find a ciphersuite spec in some BCP how-to, and want to paste it into their application settings. If a single uniform syntax supports all the relevant parameters, then all applications will expose a single parameter, which the user can set.  If they need to understand what's a signature algorithm, what's a bulk cipher and PRF, ... and find the right instance of that in each application, I think that's too much to expect.

So I am concerned that separating these settings may not be in the users' interest...

mattcaswell commented:
>  If they need to understand what's a signature algorithm, what's a bulk cipher and PRF, ... and find the right instance of that in each application, I think that's too much to expect.

If they do not understand these things then we should not be encouraging them to configure it, or copy-and-paste settings into their config. Lets not provide them with even more complex tools to play with things they don't understand. It is our responsibility to provide good defaults that "just work" for most people. If people *do* understand what these settings do then we provide them with the capability to do it. If they understand it then they should have no problem with the current approach (it will fit their mental model).

> So I am concerned that separating these settings may not be in the users' interest...

I see it the other way around. Combining them is not in their interest.

tmshort commented:
I was always annoyed at the limitations of TLSv1.2 and earlier, in that everything was included in the cipher suite definitions and nothing was orthogonal. Configuring TLSv1.3 is no different than ordering off a menu; pick something from column A, B and C. It's not that hard.

It is certainly within an applications domain to choose how it presents configuration options. It can combine them into limited options for its users, or allow orthogonal flexibility.

As a library, OpenSSL should provide the flexibility that applications need in order to create their interfaces; we should not impose restrictions on them.

IMHO, the proposed mechanism of separating provides the flexibility that applications need.

mattcaswell commented:
In spite of the discussions and disagreements on long term approach, I don't think anyone is actually objecting to *this* PR. So will someone approve it?

t-j-h commented:
I actually think we should be providing configuration options like those for TLS-1.3 for the pre-TLS-1.3 concepts and educate people about a way of setting each concept separately rather than as the traditional everything-jammed-into-one ciphers string.

t-j-h commented:
BTW I've read through the code and I'm okay to approve it - but I'll hold off for a day in case someone actually wants to take this to a vote - i.e. @vdukhovni if you strongly disagree please put in a -1 so we can take it to a vote. 

My suggestion above about having early TLS versions support separate configuration would of course be a separate PR ...

mattcaswell commented:
> I actually think we should be providing configuration options like those for TLS-1.3 for the pre-TLS-1.3 concepts and educate people about a way of setting each concept separately rather than as the traditional everything-jammed-into-one ciphers string.

I think that sounds like the kind of thing @kroeckx was suggesting above.

mattcaswell commented:
I rebased this to resolve a conflict with master. The rebase caused a test to start failing in ssapitest, so I also fixed that. While doing those two things I also found a spot I missed in clienthellotest in the conversion to the new TLSv1.3 ciphersuite configuration - so I added a commit to fix that too.

mattcaswell commented:
Rebased again to resolve more conflicts. Ping @t-j-h?

t-j-h commented on doc/man1/ciphers.pod:
```diff
@@ -705,11 +717,11 @@ Note: the CBC modes mentioned in this RFC are not supported.
+ TLS_CHACHA20_POLY1305_SHA256               TLS_CHACHA20_POLY1305_SHA256
+ TLS_AES_128_CCM_SHA256                     TLS_AES_128_CCM_SHA256
+ TLS_AES_128_CCM_8_SHA256                   TLS_AES_128_CCM_8_SHA256
 
```
I don't get this - a table of two columns which are identical ...

mattcaswell replied:
This follows on in the same format as for all of the other ciphersuites listed just before this section. The first column is the standard RFC name. The second column is the OpenSSL name. As I mentioned in the initial commentary for this PR:

> I have taken the opportunity to replace the OpenSSL specific names for the ciphersuites with the RFC standard names. The only reason to have OpenSSL specific ones was to make them a bit more consistent with the rest of the ciphersuite names. Now that they are treated differently anyway that reason seems less compelling (if it ever was).

t-j-h commented on ssl/ssl_lib.c:
```diff
@@ -2915,7 +3018,15 @@ SSL_CTX *SSL_CTX_new(const SSL_METHOD *meth)
+            "TLS_AES_256_GCM_SHA384:"
+            "TLS_CHACHA20_POLY1305_SHA256:"
+            "TLS_AES_128_GCM_SHA256"))
+        goto err;
```
That default string should be coming out of a define in ssl.h - i.e. TLS_DEFAULT_CIPHERSUITES to match SSL_DEFAULT_CIPHER_LIST ?

mattcaswell replied:
Good idea. Done.

t-j-h commented:
I've done a pass through - just a couple of minor comments.
I also think we should allow the TLS_ prefix on the ciphersuites to be optional - i.e. work without that being specified. It just really seems out of place to force its usage (and it is redundant in the context where used).

I have no issue with the strings being there - just suggesting we should also allow the non-"TLS_" versions to work.

Other than the noted comments and the suggestion above, I think this is ready.

mattcaswell commented:
> I also think we should allow the TLS_ prefix on the ciphersuites to be optional - i.e. work without that being specified. It just really seems out of place to force its usage (and it is redundant in the context where used).
>
> I have no issue with the strings being there - just suggesting we should also allow the non-"TLS_" versions to work.

I disagree with this. Having "TLS_" on the front doesn't seem a huge hardship and I think having two names for the same thing is needlessly confusing (it would not necessarily be obvious to most users that TLS_AES_256_GCM_SHA384 and AES_256_GCM_SHA384 are in fact the same thing).

mattcaswell commented:
I have responded to all the latest comments above, and have pushed a new commit to address one of them.

mattcaswell commented:
Pushed. Thanks!

tmshort commented:
Would it make any sense to have the default cipher strings be the return value of a function? Perhaps not now, but in 1.2?
